### DISCUSSION TITLE
Query on out() edges also evaluates in() edges?

### DISCUSSION THREAD

**Discussion by TitiHl** [JanusGraph/janusgraph#1309] [Q&A]
This cost long wait where the vertex has lots of incoming edges but no outgoing edges, and when I query on out(), it still evaluates incoming edges? 

```
gremlin> g.V(14590024).out().profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[14590024])                                           1           1           0.053     0.00
JanusGraphVertexStep(OUT,vertex)                                                            3591.024   100.00
    \_condition=(EDGE AND visibility:normal)
    \_isFitted=false
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@801a60ee
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                 0.024
  backend-query                                                   158522                    3388.342
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@801a60ee
                                            >TOTAL                     -           -        3591.078        -
gremlin> g.V(14590024).out().count()
==>0
gremlin> g.V(14590024).in().count()
==>158522
gremlin> g.V(14590024).outE().inV().count() . <--- still takes long time
==>0
gremlin>
```

However, if I add conditions to filter by edge label, it doesn't evaluate all edges:
```
gremlin> g.V(14590024).outE("references").inV().profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[14590024])                                           1           1          22.902    54.20
JanusGraphVertexStep(OUT,[references],vertex)                                                 19.349    45.80
    \_condition=type[references]
    \_isFitted=true
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81b80cdb
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                 0.177
  backend-query                                                        0                      18.876
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81b80cdb
                                            >TOTAL                     -           -          42.251        -
gremlin>
``` 
Looks like a bug or I miss Sth. here?

Thanks,
Alex

**Comment by TitiHl** [comment]
Wondering is this something related to the edge index? as all my edges has labeled. as I see the second query says `isFitted=true`

**Comment by farodin91** [comment]
Janusgraph have to went overall edges independent from the direction. This issue can't be easily changed in the moment, see https://docs.janusgraph.org/latest/data-model.html. 
@FlorianHockmann Correct or?

**Comment by TitiHl** [comment]
Hmmm. this doesn't explain it doesn't go through `in` edges when I put an edge conditions, i.e.
`outE("references")`

**Comment by FlorianHockmann** [comment]
What @farodin91 meant was that according to that article in the docs, all edges are just stored as an adjacency list in one row. The section [_Individual Edge Layout_](https://docs.janusgraph.org/0.3.1/data-model.html#_individual_edge_layout) shows that the label id is stored together with the edge direction in one cell which is used for sorting the edges in the adjacency list. That would explain why you see a better performance when you specify the edge direction together with the edge label as you do in your 2nd example compared to the scenario where you just specify the direction.
So, if JanusGraph only has the direction available, then it has to iterate over all edges.

**Comment by TitiHl** [comment]
Thanks @FlorianHockmann and @farodin91 for the explanation, it just feels confusing when calling inE() and outE() are actually the same as bothE().

So what would be the best practical solution for this as I can image it is a quite common scenario among quires. Approach we tried:
1. just outE(edgeLabel1, EdgeLabel2...), and we have around 10 edge labels and turns out still not efficiency enough, but way better than just calling outE().
2. move the edgeLabel to properties, and only labelling edge as `in` and `out`, so that when I call outE('out`) it only quires outgoing edges. but lost the benefits of edge labelling/indexing.

What is your recommendation of doing this in an efficient way? thanks.
